### PR TITLE
ft-add-department-teams: add departmental teams

### PR DESCRIPTION
**Pull request by gitau-kinuthia**
#### What does this PR do?
Add functionality for adding departmental teams and assigning assets to them

#### Description of Task to be completed?
- add teams model
- allow users to be  assigned to teams
- add functionality to allow assignments of assets to teams
- add functionality to view assets assigned to a team

#### How should this be manually tested?
1. On browser, goto url `http://127.0.0.1:8000/api/v1/docs/live/` to access all endpoints
2. Scroll to the `department-teams` section to view all associated endpoints for CRUD
`http://127.0.0.1:8000/api/v1/department-teams/`
`http://127.0.0.1:8000/api/v1/department-teams/<id>`

#### What are the relevant pivotal tracker stories?
[#167427843](https://www.pivotaltracker.com/story/show/167427843)

#### Screenshots (If applicable)
image

image

image

### CODE REVIEW

File: api/tests/test_departmental_teams_api.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    @patch("api.authentication.auth.verify_id_token")
+    def test_authenticated_user_get_department_teams(self, mock_verify_token):
+        mock_verify_token.return_value = {"email": self.admin_user.email}
+        response = client.get(
+            self.departmental_teams_url,
+            HTTP_AUTHORIZATION="Token {}".format(self.token_user),
+        )
+        data = response.json()
+        assert isinstance(data['results'], list)
+        assert response.status_code == status.HTTP_200_OK
```
Comment (kmazi) [inline api/tests/test_departmental_teams_api.py]: It's ok to let at least one of your test method show all the dictionary key and value expected from hitting the `departmental_teams_url` endpoin in the assertions.. Then other ones could show only specific assertions particular to the test method.

File: api/tests/test_departmental_teams_api.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+    @patch("api.authentication.auth.verify_id_token")
+    def test_authenticated_user_get_specific_department_teams_no_data(
```
Comment (kmazi) [inline api/tests/test_departmental_teams_api.py]: What does `no_data` in the test method name mean?

File: api/tests/test_departmental_teams_api.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        self, mock_verify_token
+    ):
+        mock_verify_token.return_value = {"email": self.admin_user.email}
+        url = "{}/{}/".format(self.departmental_teams_url, self.departmental_team_1.id)
+        response = client.get(
+            url, HTTP_AUTHORIZATION="Token {}".format(self.token_user)
+        )
+        data = response.json()
+        assert response.status_code == status.HTTP_200_OK
+        assert "assets_assigned" in data
```
Comment (kmazi) [inline api/tests/test_departmental_teams_api.py]: Your assertions could show the expected fields when querying for a specific department team just like it should when querying for all teams of a department.

File: api/tests/test_departmental_teams_api.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        allocation = AllocationHistory.objects.create(
+            asset=new_asset, current_assignee=self.asset_assignee_team
+        )
+        # authenticate a user
+        mock_verify_token.return_value = {"email": self.admin_user.email}
+        url = "{}/{}/".format(self.departmental_teams_url, self.departmental_team.id)
+        team_response = client.get(
+            url, HTTP_AUTHORIZATION="Token {}".format(self.token_user)
+        )
+        asset = AssetSerializer(new_asset).data
```
Comment (kmazi) [inline api/tests/test_departmental_teams_api.py]: This is unnecessary I suppose; We could do without it.

File: api/tests/test_departmental_teams_api.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+            "description": "string",
+            "department": self.department.id,
+        }
+        response = client.post(
+            self.departmental_teams_url,
+            data=payload,
+            HTTP_AUTHORIZATION="Token {}".format(self.token_user),
+        )
+        data = response.json()
+        assert isinstance(data, dict)
```
Comment (kmazi) [inline api/tests/test_departmental_teams_api.py]: What keys are returned after a post request? You test could show that.

File: settings/base.py [hunk 6]
```diff
@@ -33,16 +33,15 @@
...
 # See https://docs.djangoproject.com/en/2.0/howto/deployment/checklist/
 
 # SECURITY WARNING: keep the secret key used in production secret!
 
 SECRET_KEY = config("SECRET_KEY")
 
-DATABASES = {"default": dj_database_url.config()}
-
+# DATABASES = {"default": dj_database_url.config()}
+DATABASES = {"default": dj_database_url.config(default=config('DATABASE_URL'))}
```
Comment (kmazi) [inline settings/base.py]: Why is this necessary?

File: core/models/centre.py [hunk 7]
```diff
@@ -167,3 +167,15 @@ class Meta:
 
     def __str__(self):
         return self.name
+
+
+class DepartmentalTeam(models.Model):
```
Comment (kmazi) [inline core/models/centre.py:172]: There seem to be no tests for the expected behavior of the `DepartmentalTeam` model.

File: api/tests/__init__.py [hunk 8]
```diff
@@ -240,6 +249,7 @@ def setUpClass(cls):
         cls.centre_url = reverse("andela-centres-list")
         cls.country_url = reverse("countries-list")
         cls.department_url = reverse("departments-list")
+        cls.departmental_teams_url = reverse("departmental-teams-list")
```
Comment (kmazi) [inline api/tests/__init__.py]: This is used only in the DepartmentTeamTestCase. We could move it there. Cleaning up this setup class can help improve the run time of tests significantly.

File: api/tests/__init__.py [hunk 9]
```diff
@@ -171,6 +177,9 @@ def setUpClass(cls):
         cls.asset_assignee = apps.get_model("core", "AssetAssignee").objects.get(
             user=cls.user
         )
+        cls.asset_assignee_team = apps.get_model(
```
Comment (kmazi) [inline api/tests/__init__.py]: This can move also to the DepartmentTeamTestCase `setUp` method.

File: api/tests/__init__.py [hunk 10]
```diff
@@ -59,6 +59,12 @@ def setUpClass(cls):
         cls.department_travel = apps.get_model("core", "Department").objects.create(
             name="Travel"
         )
+        cls.departmental_team = apps.get_model(
```
Comment (kmazi) [inline api/tests/__init__.py]: Both can move to the DepartmentTeamTestCase setUp method.

File: api/serializers/andela_centres.py [hunk 11]
```diff
@@ -196,3 +199,56 @@ class CountrySerializer(serializers.ModelSerializer):
...
+    def get_assets_assigned(self, obj):
+        """
+        list all assets assigned to  a specific team
+        :param obj:
+        :return:
+        """
+
+        team_assignee = models.AssetAssignee.objects.filter(team=obj).first()
+        # The following condition introduced to make sure that results are not returned if there are orphan
+        # records in the that have no assignee (assigned_to field is optional so its possible to have an empty
```
Comment (kmazi) [inline api/serializers/andela_centres.py]: Good doc though with slight typo error.

File: api/serializers/andela_centres.py [hunk 12]
```diff
@@ -196,3 +199,56 @@ class CountrySerializer(serializers.ModelSerializer):
...
+        if team_assignee:
+            assets = models.Asset.objects.filter(assigned_to=team_assignee)
+            page = self.context["view"].paginate_queryset(assets)
+            serialized_assets = DepartmentAssetSerializer(page, many=True)
+            paginated_assets = self.context["view"].get_paginated_response(
+                serialized_assets.data
+            )
+            return paginated_assets.data
+        # return a standardised response to the api with the same structure as when there are results
+        return {"count": 0, "next": None, "previous": None, "results": []}
```
Comment (kmazi) [inline api/serializers/andela_centres.py:239]: Nice work.

File: .circleci/config.yml [hunk 13]
```diff
@@ -98,6 +98,11 @@ jobs:
       - *cmd_restore_cache
       - attach_workspace:
           at: ~/art-backend/tmp
+      - run:
+          name: run migrations
+          command: |
+            pipenv run python manage.py makemigrations
```
Comment (kmazi) [inline .circleci/config.yml]: Why make migrations on circle ci again? All you need to do is run migration that has been made locally.

**Review by kmazi — CHANGES_REQUESTED** [review]
You could also show in your test that when querying the user endpoint, the response also contains the team the user belongs to.

File: .circleci/config.yml (hunk 13 above)
Comment (gitau-kinuthia) [reply .circleci/config.yml]: Absence of this was the error that was causing my tests to fail on Circle CI. Once I added this line of code my tests could pass without throwing psycopg2 errors.

Reference : https://stackoverflow.com/questions/48184300/when-in-run-test-cases-then-i-will-get-this-error-psycopg2-operationalerror-cu?rq=1

File: api/tests/test_departmental_teams_api.py (hunk 2 above)
Comment (gitau-kinuthia) [reply api/tests/test_departmental_teams_api.py]: The test has been renamed to `test_authenticated_user_get_specific_department_teams_with_no_assets_assigned` to be more communicative

File: core/models/centre.py (hunk 7 above)
Comment (gitau-kinuthia) [reply core/models/centre.py:172]: This have been added

File: api/tests/test_departmental_teams_api.py (hunk 4 above)
Comment (gitau-kinuthia) [reply api/tests/test_departmental_teams_api.py]: This test case is necessary to show the team_name field exists in the assets model. I have refactored this to a separate test to ensure test autonomy

Comment (kmazi) [reply api/tests/test_departmental_teams_api.py]: Ok. I was referring to calling the serializer here. We could just read off the new_asset data directly for assertion.

File: api/tests/test_departmental_teams_api.py [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+            HTTP_AUTHORIZATION="Token {}".format(self.token_user),
+        )
+        data = response.json()
+        # return types assertions
+        assert isinstance(data['results'], list)
+        assert isinstance(data['results'][0], dict)
+        # results field assertions
+        assert 'name' in data['results'][0]
+        assert 'description' in data['results'][0]
+        assert 'department' in data['results'][0]
```
Comment (kmazi) [inline api/tests/test_departmental_teams_api.py]: A better assertion would be `self.assertEqual(data["results"][0]["department"], self.department_team.department)`. Let's assert for the values of those keys rather than the keys. We can do this for the other tests.

File: api/tests/test_departmental_teams_api.py [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+    @patch("api.authentication.auth.verify_id_token")
+    def test_authenticated_user_get_department_teams(self, mock_verify_token):
+        mock_verify_token.return_value = {"email": self.admin_user.email}
+        response = client.get(
+            self.departmental_teams_url,
+            HTTP_AUTHORIZATION="Token {}".format(self.token_user),
+        )
+        data = response.json()
+        # return types assertions
+        assert isinstance(data['results'], list)
```
Comment (kmazi) [inline api/tests/test_departmental_teams_api.py]: This is pytest assertion style. Would this run if the test is run with the `python manage.py test` command?

Comment (kmazi) [reply api/tests/test_departmental_teams_api.py]: Let's make the changes so that both `pytest` command and `python manage.py test` command works for all tests.

File: api/tests/test_departmental_teams_api.py [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+        )
+        mock_verify_token.return_value = {"email": self.admin_user.email}
+        url = "{}/{}/".format(self.departmental_teams_url, self.departmental_team.id)
+        team_response = client.get(
+            url, HTTP_AUTHORIZATION="Token {}".format(self.token_user)
+        )
+        data = team_response.json()
+        assert "members" in data
+        assert len(data["members"]["results"]) == 1
+        assert data["members"]["results"][0]["email"] == new_user.email
```
Comment (kmazi) [inline api/tests/test_departmental_teams_api.py]: Great!! These assertions are super cool.

File: core/tests/test_department_team.py [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+            name="Accounts", department_id=self.new_department.id
+        )
+        new_team.save()
+        assert DepartmentalTeam.objects.count() == count + 1
+        new_team.delete()
+
+    def test_cannot_add_existing_team(self):
+        team = DepartmentalTeam(name="Accounts", department_id=self.new_department.id)
+        team.save()
+        with transaction.atomic():
```
Comment (kmazi) [inline core/tests/test_department_team.py]: The default TestCase wraps all tests methods in a transaction. There is no real need to explicitly manage the transactions here in the tests.

File: core/tests/test_department_team.py [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+            with pytest.raises(utils.IntegrityError):
```
Comment (kmazi) [inline core/tests/test_department_team.py]: This also is germane to pytest. Would this work with `python manage.py test`? Unless we're planning to deprecate running with django test.

**Review by kmazi — CHANGES_REQUESTED** [review]

File: core/tests/test_department_team.py (hunk 17 above)
Comment (gitau-kinuthia) [reply core/tests/test_department_team.py]: @kmazi the atomic block is actually necessary. In its absence, the following error is obtained:
```django.db.transaction.TransactionManagementError: An error occurred in the current transaction. You can't execute queries until the end of the 'atomic' block.```

File: core/tests/test_department_team.py (hunk 18 above)
Comment (gitau-kinuthia) [reply core/tests/test_department_team.py]: Yes the tests still run using the ```python manage.py test```

File: api/tests/test_departmental_teams_api.py (hunk 15 above)
Comment (gitau-kinuthia) [reply api/tests/test_departmental_teams_api.py]: > This is pytest assertion style. Would this run if the test is run with the `python manage.py test` command?

Yes

File: core/tests/test_department_team.py (hunk 17 above)
Comment (kmazi) [reply core/tests/test_department_team.py]: My point is that we don't need a db transaction for this test method especially because the TestCase base class handles transaction for all test methods. Therefore calling it explicitly is unnecessary. @fgkinus

File: core/tests/test_department_team.py (hunk 18 above)
Comment (gitau-kinuthia) [reply core/tests/test_department_team.py]: This has been refactored however to the unittest assertion  style

File: api/tests/test_departmental_teams_api.py (hunk 15 above)
Comment (gitau-kinuthia) [reply api/tests/test_departmental_teams_api.py]: > Let's make the changes so that both `pytest` command and `python manage.py test` command works for all tests.

This has been refactored however to the unittest assertion  style

File: api/tests/test_departmental_teams_api.py (hunk 14 above)
Comment (gitau-kinuthia) [reply api/tests/test_departmental_teams_api.py]: This has been refactored however to the unittest assertion  style

File: core/tests/test_department_team.py (hunk 17 above)
Comment (gitau-kinuthia) [reply core/tests/test_department_team.py]: Noted. I have been able to remove the delete function calls that were eliciting the transaction errors.

**Review by kmazi — APPROVED** [review]
